woodwork/schema.py

```diff
@@ -154,27 +154,46 @@ def time_index(self):
-        """Update the semantic tags for any columns names in the provided types dictionary,
-        updating the Woodwork typing information for the DataFrame.
+    def set_types(self, logical_types=None, semantic_tags=None, retain_index_tags=True):
+        """Update the logical type and semantic tags for any columns names in the provided types dictionaries,
+        updating the Schema at those columns.
 
         Args:
+            logical_types (dict[str -> LogicalType], optional): A dictionary defining the new logical types for the
+                specified columns.
             semantic_tags (dict[str -> str/list/set], optional): A dictionary defining the new semantic_tags for the
                 specified columns.
             retain_index_tags (bool, optional): If True, will retain any index or time_index
-                semantic tags set on the column. If False, will replace all semantic tags. Defaults to
-                True.
-
+                semantic tags set on the column. If False, will replace all semantic tags any time a column's
+                semantic tags or logical type changes. Defaults to True.
         """
+        logical_types = logical_types or {}
+        _check_logical_types(self.columns.keys(), logical_types, require_all_cols=False)
+
         semantic_tags = semantic_tags or {}
         _check_semantic_tags(self.columns.keys(), semantic_tags)
 
-        for col_name, new_tags in semantic_tags.items():
+        for col_name in (logical_types.keys() | semantic_tags.keys()):
             original_tags = self.semantic_tags[col_name]
-            new_semantic_tags = _set_semantic_tags(new_tags,
-                                                   self.logical_types[col_name].standard_tags,
-                                                   self.use_standard_tags)
-            _validate_not_setting_index_tags(new_semantic_tags, col_name)
+
+            # Update Logical Type for the Schema, getting new semantic tags
+            ltype_update_tags = original_tags
+            new_logical_type = logical_types.get(col_name)
+            if new_logical_type is not None:
+                self.columns[col_name]['logical_type'] = new_logical_type
+                ltype_update_tags = _reset_semantic_tags(new_logical_type.standard_tags, self.use_standard_tags)
+
+            # Get new semantic tags, removing existing tags
+            new_semantic_tags = semantic_tags.get(col_name)
+            if new_semantic_tags is not None:
+                new_semantic_tags = _set_semantic_tags(new_semantic_tags,
+                                                       self.logical_types[col_name].standard_tags,
+                                                       self.use_standard_tags)
+                _validate_not_setting_index_tags(new_semantic_tags, col_name)
+
+            # Update the tags for the Schema
+            if new_semantic_tags is None:
+                new_semantic_tags = ltype_update_tags
```

Comment:
I think we still need some intermediate like `ltype_update_tags` so that `new_semantic_tags` can take precedence when it's present, but if it's `None`, it'll mean that there was an ltype update, so we need `ltype_update_tags` to reset the semantic tags`

Reply:
Not sure if it is better or not, but this is what I was thinking for the `for` loop:
```python
        for col_name in (logical_types.keys() | semantic_tags.keys()):
            original_tags = self.semantic_tags[col_name]

            # Update Logical Type for the Schema, getting new semantic tags
            new_logical_type = logical_types.get(col_name)
            if new_logical_type is not None:
                self.columns[col_name]['logical_type'] = new_logical_type
                # ltype_update_tags = _reset_semantic_tags(new_logical_type.standard_tags, self.use_standard_tags)

            # Get new semantic tags, removing existing tags
            new_semantic_tags = semantic_tags.get(col_name)
            if new_semantic_tags is not None:
                new_semantic_tags = _set_semantic_tags(new_semantic_tags,
                                                       self.logical_types[col_name].standard_tags,
                                                       self.use_standard_tags)
                _validate_not_setting_index_tags(new_semantic_tags, col_name)
            else:
                new_semantic_tags = _reset_semantic_tags(new_logical_type.standard_tags, self.use_standard_tags)

            self.columns[col_name]['semantic_tags'] = new_semantic_tags

            if retain_index_tags and 'index' in original_tags:
                self._set_index_tags(col_name)
            if retain_index_tags and 'time_index' in original_tags:
                self._set_time_index_tags(col_name)
```

Reply:
I think that works because we know if `new_semantic_tags` is `None` that means `new_logical_type` is not `None`.

Reply:
Ahh, got it. That'll definitely work, but I think it might be beneficial to keep the actual reset call next to where we update the logical type for clarity.

Maybe if we do `if new_semantic_tags is None` then `new_semantic_tags = _reset_semantic_tags(..)` then we can have the concept of "if we change the logical type the tags get reset" more closely tied together

Reply:
Yep, sounds good!

Reply:
As a side note here, related to a prior conversation, I think some of this confusion stems from having a method that is responsible for doing multiple things - if we had this logic completely separated into two methods this wouldn't be quite as confusing.

Reply:
With DataTables, I think we started out with separate methods and combined them to a single `set_types`... maybe @gsheni remembers why??